testsuite/test-runner/src/main/java/org/wildfly/core/testrunner/WildFlyRunner.java

```diff
@@ -184,34 +184,50 @@ public void testRunFinished(Result result) throws Exception {
             Security.insertProviderAt(ELYTRON_PROVIDER, 0);
             providerInstalled = true;
         }
+        RuntimeException setupException = null;
         try {
             if (automaticServerControl) {
-                runSetupTasks();
+                setupException = runSetupTasks();
             }
         } catch (AssumptionViolatedException e) {
             assumptionFailedInServerSetup = e;
         }
-        super.run(notifier);
+        if (setupException == null) {
+            super.run(notifier);
+        }
 
         if (automaticServerControl) {
-            runTearDownTasks();
+            RuntimeException tearDownException = null;
+            try {
+                runTearDownTasks();
+            } catch (RuntimeException e) {
+                tearDownException = e;
+            } finally {
+                if (setupException != null) {
+                    throw setupException;
+                }
+                if (tearDownException != null) {
+                    throw tearDownException;
+                }
+            }
         }
         if (providerInstalled) {
             Security.removeProvider(ELYTRON_PROVIDER.getName());
         }
     }
 
-    private void runSetupTasks() {
+    private RuntimeException runSetupTasks() {
         for (ServerSetupTask task : serverSetupTasks) {
             try {
                 task.setup(controller.getClient());
                 runSetupTasks++;
             } catch (AssumptionViolatedException e) {
                 throw e;
             } catch (Exception e) {
-                throw new RuntimeException(String.format("Could not run setup task '%s'", task), e);
+                return new RuntimeException(String.format("Could not run setup task '%s'", task), e);
             }
         }
+        return null;
```

Comment:
Just out of curiosity, why not continue to control the behaviour by throwing the RuntimeException instead of returning it as a value? At first glance, it looked more natural to me thrown an exception instead of returning it as a value.

Reply:
thanks, I think it makes sense to throw the exception , will update it. :)